File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.py
import heapq
from collections import deque
def solution(jobs):
    answer = 0
    jobs.sort(key = lambda x: [x[0],x[1]])
    job_queue = deque(jobs)
    waiting_queue = []
    heapq.heapify(waiting_queue)
    job_num = 0
    wt = 0
    #첫 작업 다루기
    first_job = job_queue.popleft()
    time = first_job[0]
    heapq.heappush(waiting_queue,[first_job[1],first_job[0],job_num])
    job_num += 1
    
    while job_queue or waiting_queue:
        # 작업 처리
        if waiting_queue :
            job = heapq.heappop(waiting_queue)
            # 시간 갱신
            if time < job[1] :
                time = job[1] + job[0]
            else :
                time += job[0]
            #대기 시간 갱신
            wt += (time - job[1])
            
        else :
            time = job_queue[0][0]
            
        # 작업큐에 들어 올 수 있는 작업들 추가
        while job_queue and job_queue[0][0] <= time :
            new_job = job_queue.popleft()
            heapq.heappush(waiting_queue,[new_job[1],new_job[0],job_num])
            job_num += 1
        
    #print(wt)
    answer = wt // len(jobs)
    return answer

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.java
import java.util.*;
class Solution {
    public int solution(int[][] jobs) {
        int N = jobs.length;
        int answer = 0;
        PriorityQueue<Job> pq = new PriorityQueue<>();
        List<Job> queue = new LinkedList<>();
        
        for(int i = 0; i< N; i++)
            queue.add(new Job(i,jobs[i][1],jobs[i][0]));
            
        Collections.sort(queue,(o1,o2) -> o1.requestTime - o2.requestTime);
        
        
        int time = 0;
        while(!queue.isEmpty() || !pq.isEmpty()){
            //작업 큐에 들어올 수 있는 작업들
            while(!queue.isEmpty() && queue.get(0).requestTime <= time){
                Job job = queue.remove(0);
                pq.add(job);
                
            }
            
            if(!pq.isEmpty()){
                Job job = pq.poll();
                if(time < job.requestTime)
                    time = job.requestTime;
                
                
                time += job.duration;
                answer += (time - job.requestTime);
            }
            else if(!queue.isEmpty()){
                time = queue.get(0).requestTime;
            }
        }
        
        
        return answer / N;
    }
}
class Job implements Comparable<Job>{
    int num;
    int duration;
    int requestTime;
    
    public Job(int num, int duration, int requestTime){
        this.num = num;
        this.duration = duration;
        this.requestTime = requestTime;
    }
    
    @Override
    public int compareTo(Job o){
        if(this.duration == o.duration){
            if(this.requestTime == o.requestTime){
                return this.num = o.num;
            }
            else return this.requestTime - o.requestTime;
        }
        else return this.duration - o.duration;
    }
    
}